File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.cc
//백준 12865, 평범한 배낭

#include <iostream>

int w[105];
int v[105];
int dp[105][100'005];

int main (){

    int N, K;
    std::cin >> N >> K;
    for(int i{1}; i<=N; ++i){
        std::cin >> w[i] >> v[i];
    }

    for(int i{1}; i<=N; ++i){
        for(int j{1}; j<=K; ++j){
            if(w[i] > j){
                dp[i][j] = dp[i-1][j];
            }
            else{
                dp[i][j] = std::max(dp[i-1][j-w[i]] + v[i], dp[i-1][j]);
            }
        }
    }

    std::cout << dp[N][K];

    return 0;
}

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
N, K = map(int, input().split())

dp = [[0]*(K+1) for _ in range(N+1)]

W = [0]
V = [0]

for _ in range(N):
  w, v = map(int, input().split())
  W.append(w); V.append(v)

for i in range(1, K+1):
  for j in range(1, N+1):
    if i >= W[j]:
      dp[j][i] = max(dp[j-1][i], dp[j-1][i-W[j]]+V[j])
    else:
      dp[j][i] = dp[j-1][i]

print(dp[N][K])